### ISSUE TITLE
Preview of FILTER RULE implementation

### ISSUE THREAD

**Issue by usrflo** [perlundq/yajsync#18]
Please take a look at https://github.com/usrflo/yajsync/commit/378754a72c0ce5bbffe50f06171ce52c194321ae
These classes implement http://rsync.samba.org/ftp/rsync/rsync.html, section FILTER RULES, INCLUDE/EXCLUDE PATTERN RULES and MERGE-FILE FILTER RULES.

I could integrate this filter rule processing in yajsync *if you don't have other plans*. The integration consists of:
- reading option arguments
- initialize the filter rule configuration classes according to the rsync filter rule syntax
- check the rules on dir/file list processing

Please tell me if you are interested.

**Comment by usrflo** [comment]
I increased the compatibility of FILTER RULES in https://github.com/usrflo/yajsync/tree/master to rsync. The system tests consider yajsync-client or rsync-client to yajsync-server. Core support is available but more tests have to be written to cover more cases. Considering the current state of stability the latter should be part of future development on a consolidated code basis (?)

The changes for FILTER RULES (-f, --filter, --exclude, --exclude-from, --include, --include-from, --delete, --delete-excluded) in https://github.com/usrflo/yajsync/tree/master contain changes to "preserve group" (-g, --group) and "numeric-ids" (--numeric-ids). I think it would be best I extract this to 3 different pull requests based on your latest checkin. May it be better to wait for a new release from your side to decrease merge conflicts?

**Comment by perlundq** [comment]
Yes split it into three separate patches. preserve group and numeric ids should be fairly easy to merge. If the filtering rules work as expected that will go in too. Please try to adhere to the coding guidelines, mainly for the filtering rules, (not a big deal, but I will fix that myself otherwise)

**Comment by usrflo** [comment]
I now manually merged my changes from 2014, 2015 and 2016 to the following branch: https://github.com/usrflo/yajsync/tree/filter2. This branch is based on your last master branch. The SystemTests are running but I think the remote modes should get better test coverage in the next step (I am quite unsure with correct filter modes for sender/receiver/generator in RsyncClient.java).

There are some things to clarify before I want to request for a merge:

- the current system tests mainly focus on local mode: wouldn't it be better to enable the same test cases for local, remote and native rsync test runs like I proposed in my 'systemtest' branch? There may be better implementations than abstract methods but these modes should be simply exchangeable to get a better test coverage.

- you added the option --delete inbetween with deletion in the generator. I assumed the deletion has to be executed from within the receiver, see my implementation. Deletion methods seem to be implemented twice now, so how can this be consolidated?

**Comment by perlundq** [comment]
I haven't had the time to look into this yet. Regarding the tests I think it would be better (at least to start with) to have a few explicit tests for compatibility with native rsync (I am a bit unsure, but it might be cleaner to use junit parameterized with rsync and yajsync instead, if you see what I mean).

Regarding the delete, I suggest adapting to the already present implementation (--delete must not every be buggy and I have put some thought into that one already)

**Comment by usrflo** [comment]
https://github.com/usrflo/yajsync/tree/filter2 now contains SystemTest with parameterized yajsync/rsync tests. This setup is a bit more complicated than the abstract test class approach. Moreover a parameter dependent skipping of tests has to be done explicitly by a instance check of the test class factory. Anyway this approach is working.

Interesting to see that for rsync e.g. com.github.perlundq.yajsync.test.SystemTest.testClientDirCopyDstDeletion() behaves differently to yajsync: the thrown (but not catched) exception in com.github.perlundq.yajsync.session.Receiver.receiveFiles(Filelist, Segment) is:
                        throw new IllegalStateException(
                            "got file list DONE with empty file list and at " +
                            "least all ouststanding segment deletions " +
                            "acknowledged but haven't received file list EOF");

In this case the receiver has to be re-checked (plus the exception handling should be improved: at least there should be a log statement in case of IllegalStateExceptions). In other failing test cases in SystemTest the filter implementation has to be reviewed for native rsync compatibility.

**Comment by usrflo** [comment]
Please re-open this issue so the filter topics gets more attention.

**Comment by usrflo** [comment]
@perlundq : could you please review https://github.com/usrflo/yajsync/tree/filter2 and give feedback to

- the implementation of parameterized system tests (for parallel yajsync and rsync testing)

- the filter implementation itself

It would be great if yajsync could be extended by filter options in 2017.

**Comment by perlundq** [comment]
OK, yes I will try to find some time in January for this

**Comment by usrflo** [comment]
@perlundq : is there any chance you take a look at the linked filter implementation?
